opentelemetry-sdk/src/metrics/periodic_reader_with_own_thread.rs

```diff
@@ -0,0 +1,798 @@
+        // TODO: This part might benefit from a different design.
+        let meter_provider = SdkMeterProvider::builder()
+            .with_reader(reader.clone())
+            .build();
+
+        // Now collect and flush should succeed
+        let result = reader.collect(rm);
+        assert!(result.is_ok());
+
+        let result = meter_provider.force_flush();
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn exporter_failures_are_handled() {
+        // create a mock exporter that fails 1st time and succeeds 2nd time
+        // Validate using this exporter that periodic reader can handle exporter failure
+        // and continue to export metrics.
+        // Arrange
+        let interval = std::time::Duration::from_millis(10);
+        let exporter = MetricExporterThatFailsOnlyOnFirst::default();
+        let reader = PeriodicReaderWithOwnThread::builder(exporter.clone())
+            .with_interval(interval)
+            .build();
+
+        let meter_provider = SdkMeterProvider::builder().with_reader(reader).build();
+        let meter = meter_provider.meter("test");
+        let counter = meter.u64_counter("sync_counter").build();
+        counter.add(1, &[]);
+        let _obs_counter = meter
+            .u64_observable_counter("testcounter")
+            .with_callback(move |observer| {
+                observer.observe(1, &[]);
+            })
+            .build();
+
+        // Sleep for a duration much longer than the interval to trigger
+        // multiple exports, including failures.
+        // Not a fan of such tests, but this seems to be the
+        // only way to test if periodic reader is doing its job. TODO: Decide if
+        // this should be ignored in CI
+        std::thread::sleep(Duration::from_millis(500));
+
+        // Assert that atleast 2 exports are attempted given the 1st one fails.
+        assert!(exporter.get_count() >= 2);
+    }
+
+    #[test]
+    fn collection() {
+        collection_triggered_by_interval_helper();
+        collection_triggered_by_flush_helper();
+        collection_triggered_by_shutdown_helper();
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn collection_from_tokio_multi_with_one_worker() {
+        collection_triggered_by_interval_helper();
+        collection_triggered_by_flush_helper();
+        collection_triggered_by_shutdown_helper();
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn collection_from_tokio_with_two_worker() {
+        collection_triggered_by_interval_helper();
+        collection_triggered_by_flush_helper();
+        collection_triggered_by_shutdown_helper();
+    }
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn collection_from_tokio_current() {
+        collection_triggered_by_interval_helper();
+        collection_triggered_by_flush_helper();
+        collection_triggered_by_shutdown_helper();
```

Comment:
nit - These 3 helper methods can be wrapped in a single method and called in all test cases to reduce redundancy